auth_matrix/admin_views.py

```diff
@@ -1,15 +1,37 @@
+from typing import Any
+
 from django import views
-from django.contrib.admin.views.decorators import staff_member_required
 from django.contrib.auth import get_user_model
+from django.contrib.auth.mixins import UserPassesTestMixin
 from django.contrib.auth.models import Group, Permission
-from django.utils.decorators import method_decorator
+from django.http import HttpRequest
+from django.http.response import HttpResponse
 
 User = get_user_model()
 
-@method_decorator(staff_member_required, name="dispatch")
-class AuthorizationMatrixView(views.generic.base.TemplateView):
 
-    template_name = "admin/authorization/authorization_matrix.html"
+class UserIsStaffMixin(UserPassesTestMixin):
+    def test_func(self):
+        return self.request.user.is_staff
+
+
+def get_auth_matrix_permission_label():
+    opts = User._meta
+    return f"{opts.app_label}.view_{opts.model_name}"
+
+
+class CanViewAuthorizationMatrixMixin(UserIsStaffMixin):
+    def dispatch(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
+        if not request.user.has_perm(get_auth_matrix_permission_label()):
+            return self.handle_no_permission()
+        return super().dispatch(request, *args, **kwargs)
```

Comment:
The `PermissionRequiredMixin` already has a hook to dynamically return/calculate the required permission: `get_permission_required`

```py
class CanViewAuthorizationMatrixMixin(UserIsStaffMixin, PermissionRequiredMixin):

    def get_permission_required(self):
        permission_required = get_auth_matrix_permission_label()
        return (permission_required,)
```

this has a couple of benefits:

1. avoiding overriding `dispatch` (since mistakes in here have high risk, security wise)
2. the order of checks: first `UserIsStaffMixin` is evaluated, and it shortcuts if the user is not logged in or staff. Next, the built-in `PermissionRequiredMixin` is checked, and we validate it against the user model permission (dynamically determined)
3. No need to override any "unhappy flow" handling

All of the above runs during the `dispatch` phase of the view, and it's all going via inheritance. Only when all the requirements are met, the actual view is executed.